### PR TITLE
feat: add support for GitLab Generic Packages

### LINKED ISSUE #2006: Feature request: Switch GitLab release uploads to use Generic Packages Repository
**Is your feature request related to a problem? Please describe.**
With GitLab [Generic Packages Repository](https://docs.gitlab.com/ee/user/packages/generic_packages/#publish-a-generic-package-by-using-cicd) being available in Core/Free, it would be a better target for uploading binaries than the currently used [Upload Project File API endpoint](https://docs.gitlab.com/ce/api/projects.html#upload-a-file).

There are multiple benefits over "project files":
* The users can browse Generic Packages separately from and in addition to Releases in the GitLab UI
* Can still be used as release assets, as all you specify there is a URL
* API access for _downloading_ binaries is not supported for "project files", but is for Generic Packages

**Describe the solution you'd like**
Replace the upload method for GitLab releases to use the Generic Packages API, making sure the Release Asset URLs are correspondingly replaced.

**Describe alternatives you've considered**
Using the HTTP uploads feature to _additionally_ upload Generic Packages, as is possible with #2002 and #2005 combined. However, it requires additional configuration by the user, and results in two uploads per binary - one for "project files" and one for Generic Packages.

**Additional context**
GitLab documentation says the following about Generic Packages:

> Version history
> 
> - Introduced in GitLab 13.5.
> - It’s deployed behind a feature flag, enabled by default.
> - It’s enabled on GitLab.com.
> - It’s able to be enabled or disabled per-project.
> - It’s recommended for production use.
> - For GitLab self-managed instances, GitLab administrators can opt to disable it. 

- https://docs.gitlab.com/ee/user/packages/generic_packages/#publish-a-generic-package-by-using-cicd

### PR DESCRIPTION
**Pull request by neomantra**
GitLab's Generic Package Repository allows uploading files
to GitLab, as an alternative to attachments.

Added the flag `use_package_registry` to the `gitlab_urls` config.
When false, the default, it will behave as before.
When true, it will publish the file to the Generic Package Registry.
Either way, the URL is bound to the release the same

Updated `go-gitlab` to version 0.52.2 which has the new
`GenericPackages.PublishPackageFile` function.

----
Fixes #2006.  

I tried to run local tests on it and it was not working (failures with `internal/pipe` and weird assembly output);  however, I think that is unrelated to this PR as I've tested this many times before without issue.  My apologies if this is not the case and the CI fails.

### CODE REVIEW

File: www/docs/scm/gitlab.md [hunk 1]
```diff
@@ -37,6 +40,14 @@ If none are set, they default to GitLab's public URLs.
     on [release](https://docs.gitlab.com/ee/user/project/releases/index.html) functionality
     and [direct asset linking](https://docs.gitlab.com/ee/user/project/releases/index.html#permanent-links-to-release-assets).
 
+## Generic Package Registry
+
+GitLab introduced the [Generic Package Registry](https://docs.gitlab.com/ee/user/packages/package_registry/index.html) in Gitlab 13.5.  Normally, `goreleaser` uploads release files as "attachments", which may have [administrative limits](https://docs.gitlab.com/ee/user/admin_area/settings/account_and_limit_settings.html).  Notably, hosted gitlab.com instances have a 10MB attachment limit which cannot be changed.  Uploading to the Generic Package Registry does not have this restriction.  To use it instead, set `use_package_registry` to `true`.
+
+```
```
Comment (caarlos0) [inline www/docs/scm/gitlab.md:47]:
```suggestion
```yaml
# .goreleaser.yml
```

File: www/docs/scm/gitlab.md [hunk 2]
```diff
@@ -37,6 +40,14 @@ If none are set, they default to GitLab's public URLs.
     on [release](https://docs.gitlab.com/ee/user/project/releases/index.html) functionality
     and [direct asset linking](https://docs.gitlab.com/ee/user/project/releases/index.html#permanent-links-to-release-assets).
 
+## Generic Package Registry
+
+GitLab introduced the [Generic Package Registry](https://docs.gitlab.com/ee/user/packages/package_registry/index.html) in Gitlab 13.5.  Normally, `goreleaser` uploads release files as "attachments", which may have [administrative limits](https://docs.gitlab.com/ee/user/admin_area/settings/account_and_limit_settings.html).  Notably, hosted gitlab.com instances have a 10MB attachment limit which cannot be changed.  Uploading to the Generic Package Registry does not have this restriction.  To use it instead, set `use_package_registry` to `true`.
```
Comment (caarlos0) [inline www/docs/scm/gitlab.md:45]:
```suggestion
GitLab introduced the [Generic Package Registry](https://docs.gitlab.com/ee/user/packages/package_registry/index.html) in Gitlab 13.5.
Normally, `goreleaser` uploads release files as "attachments", which may have [administrative limits](https://docs.gitlab.com/ee/user/admin_area/settings/account_and_limit_settings.html).  Notably, hosted gitlab.com instances have a 10MB attachment limit which cannot be changed.
Uploading to the Generic Package Registry does not have this restriction.  To use it instead, set `use_package_registry` to `true`.
```

**Review by caarlos0 — COMMENTED** [review]
looking good overall, added just a couple of comments

File: www/docs/scm/gitlab.md (hunk 2 above)
Comment (neomantra) [reply www/docs/scm/gitlab.md:45]: Sure, changing this now -- did you make (and want)  the `not`  in line 47 the color red somehow or is that just how GitHub is showing it?

Comment (neomantra) [reply www/docs/scm/gitlab.md]: OK pushed that (without anything with `not` coloring).

**Comment by neomantra** [comment]
I have rebased this PR to main and squashed the doc PR commit.   
There are less file changes now (e.g. `go.mod`) because `go-gitlab` was upgraded to `0.52.2` in  2a340e9125f417495eac191af03c880e61004344.

**Comment by caarlos0** [comment]
LGTM! Thanks for the hard work on this!

**Pull request merged**